### ISSUE TITLE
Unify `for` loops

### ISSUE THREAD

**Issue by rattrayalex** [lightscript/lightscript#4]
The [big table of for loops](http://www.lightscript.org/docs/#which-for-art-thou-romeo) is a bummer. LightScript makes one or two cases better with `from` (iterating over arrays and certain ranges) but it's one more thing to remember (actually, several more things to remember). 

Instead, I'm thinking of the following syntax: 

## Arrays

```js
for idx i, elem x in arr: 
  print(i, x)
for idx i in arr: 
  print(i)
for elem x in arr:
  print(x)
for elem x, idx i in arr:  // not sure whether to support "flipped" args
  print(i, x)
```
```js
for (let i = 0; i < arr.length; i++) {
  const x = arr[i];
  print(i, x);
}
for (let i = 0; i < arr.length; i++) {
  print(i);
}
for (let _i = 0; _i < arr.length; _i++) {
  const x = arr[_i];
  print(x);
}
for (let i = 0; i < arr.length; i++) {  // not sure whether to support "flipped" args
  const x = arr[i];
  print(i, x);
}
```

## Objects

```js
for key k, val v in obj: 
  print(k, v)
for key k in obj: 
  print(k)
for val v in obj: 
  print(v)
for val v, key k in obj: // again, not sure whether to support flipped args
  print(k, v)
```
```js
for (const k in obj) { 
  const v = obj[k];
  print(k, v);
}
for (const k in obj) { 
  print(k);
}
for (const _key in obj) { 
  const v = obj[_key];
  print(v);
}
for (const k in obj) { // again, not sure whether to support flipped args
  const v = obj[k];
  print(k, v);
}
```

I'm also considering something similar for iterables (either `for iter x of gen` or `for elem x of *gen`) but `of` is just fine for that, and they're not so common. 

Similarly, numeric ranges aren't that common in practice and `for ;;` works fine in those cases – plus, once the stdlib is built and lodash is always-available, `for elem x in range(10, 20, 2)` will be plenty easy when the overhead of array creation is acceptable. 

I'm curious for thoughts on the above syntax before I build it.

**Comment by rattrayalex** [comment]
For backwards-compatible non-confusion, the following would be a SyntaxError: 

```js
for k in obj:
  print(k)
```
and, if seeking to iterate over the keys of an object without using `key`, would either need to write:
```js
for const k in obj:
  print(k)
```
or, to update an existing variable:
```js
let k;
for now k in obj:
  print(k)
```

**Comment by wcjohnson** [comment]
- I like this proposed syntax.

- Unless there is some grammatical reason not to, I think you should support flipping the order of the indices and values. That's one less thing for coders to remember. If you don't support it, you have to remember which form to use AND which order things go in.

- Not sure how I feel about making JS `for..in` into a syntax error. People writing LS will be used to the fact that newly declared variables are implicitly `const`. Why make them explicitly type the keyword in this particular case?

If you want to force adoption of the new syntax, then why have the `for const ... in` form at all?

**Comment by rattrayalex** [comment]
Great, I'll probably support both orderings and rely on a future ESLint rule to encourage a consistent style. Thanks!

We discussed the `for...in` SyntaxError question on Gitter, but it's a good question worth having answered publicly. 

If `for k in obj` compiled to `for (const k in obj)`, people would use that instead of `for key k in obj` because it's more concise. That'd be really nice, but makes it less clear when you're iterating over an object versus an array, which is the key challenge with unified for loops in JS. It'd be tough to remember, just reading some code, whether `for thing in things` is supposed to be for objects or arrays, and similarly whether it's keys/indices or values/elements that it's doing (what I just wrote *looks* like it's elements from an array, but it's actually keys from an object). 

Separately, `let x; x = 1` is the single biggest backwards-incompatibility in LightScript – it raises a syntax error, telling you to use `now` instead. I'm very happy with that, and requiring `now` in a for loop in which you're reassigning a previously declared variable is a natural extension of that syntax.
